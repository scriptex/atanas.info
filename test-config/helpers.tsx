import { act } from 'react-dom/test-utils';
import type { FC } from 'react';
import { render, RenderResult, waitFor } from '@testing-library/react';

import * as shared from '@scripts/shared';

type ComponentType = FC<Readonly<React.ComponentProps<any>>>;

export const test = async (Component: ComponentType): Promise<RenderResult> => {
	let result!: RenderResult;

	await act(async () => {
		result = await waitFor(() => render(<Component />));
	});

	return result;
};

export const snapshotTest = (Component: ComponentType, buttonSelector?: string): void => {
	const { name } = Component;

	jest.spyOn(shared, 'formatDate');

	(shared.formatDate as jest.Mock).mockImplementation(() => 'Mock date');

	afterAll(() => {
		jest.clearAllMocks();
	});

	describe(`${name} component`, () => {
		it(`Should render the ${name} component`, async () => {
			const { asFragment, container } = await test(Component);

			expect(asFragment()).toMatchSnapshot();

			if (buttonSelector) {
				await act(async () => {
					await container.querySelectorAll<HTMLElement>(buttonSelector).forEach(button => {
						if (!button.hasAttribute('download')) {
							button.click();

							expect(asFragment()).toMatchSnapshot();
						}
					});
				});
			}
		});

		it(`Should unmount the ${name} component`, async () => {
			const { unmount, asFragment } = await test(Component);

			unmount();

			expect(asFragment()).toMatchSnapshot();
		});
	});
};

export const mockAudioContext = (): jest.Mock => {
	const mockConnect = jest.fn();

	const mockgetByteFrequencyData = jest.fn();

	const mockcreateMediaElementSource = jest.fn(() => ({
		connect: mockConnect
	}));

	const mockcreateAnalyser = jest.fn(() => ({
		connect: mockConnect,
		frequencyBinCount: [0, 1, 2],
		getByteFrequencyData: mockgetByteFrequencyData
	}));

	const mockcreateOscillator = jest.fn(() => ({
		channelCount: 2
	}));

	const mockChannelSplitterConnect = jest.fn(n => n);

	const mockcreateChannelSplitter = jest.fn(() => ({
		connect: mockChannelSplitterConnect
	}));

	const mockedAudioContext = jest.fn(() => ({
		createAnalyser: mockcreateAnalyser,
		createOscillator: mockcreateOscillator,
		createChannelSplitter: mockcreateChannelSplitter,
		createMediaElementSource: mockcreateMediaElementSource
	}));

	return mockedAudioContext;
};

export const mockFetch = (value: Record<string, string | boolean>): void => {
	global.fetch = jest.fn().mockImplementation(() =>
		Promise.resolve({
			json: () => Promise.resolve(value)
		})
	);
};
