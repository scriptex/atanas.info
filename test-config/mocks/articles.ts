import type { Asset } from 'contentful';

import { Article } from '@scripts/cms';

export const articles: Article[] = [
	{
		external: true,
		externalImage: 'https://article.one/image.png',
		index: 0,
		title: 'Article 1 title',
		url: 'https://article.one/'
	},
	{
		external: false,
		image: {
			fields: {
				file: {
					url: 'https://article.two/image.png'
				}
			}
		} as Asset,
		index: 1,
		title: 'Article 2 title',
		url: 'https://article.two/'
	}
];
