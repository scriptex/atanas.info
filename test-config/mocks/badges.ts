import { Asset } from 'contentful';

import { Badge } from '@scripts/cms';

export const badges: Badge[] = [
	{
		date: '2020-12-01T00:00+03:00',
		image: {
			fields: {
				file: {
					url: 'https://www.file1.com/image.png'
				}
			}
		} as Asset,
		index: 0,
		issuer: 'Issuer 1',
		issuerUrl: 'https://www.issuer1.com',
		title: 'Badge 1',
		url: 'https://www.url1.com'
	},
	{
		date: '2020-12-02T00:00+03:00',
		image: {
			fields: {
				file: {
					url: 'https://www.file2.com/image.png'
				}
			}
		} as Asset,
		index: 1,
		issuer: 'Issuer 2',
		issuerUrl: 'https://www.issuer2.com',
		title: 'Badge 2',
		url: 'https://www.url2.com'
	}
];
