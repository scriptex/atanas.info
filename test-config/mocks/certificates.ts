import { Certificate } from '@scripts/cms';

export const certificates: Certificate[] = [
	{
		pdf: 'https://example.com/file.pdf',
		date: '2020-11-20T00:00:00',
		index: 0,
		image: 'https://example.com/image.png',
		title: 'Certificate 1'
	}
];
