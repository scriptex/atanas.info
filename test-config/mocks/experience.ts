import { Experience } from '@scripts/cms';

export const experience: Experience[] = [
	{
		index: 0,
		title: 'Experience title 1',
		project: '<p>Experience project 1</p>',
		location: 'Location 1',
		period: 'Period start - Period end',
		details: '<p>Experience details 1</p>'
	}
];
