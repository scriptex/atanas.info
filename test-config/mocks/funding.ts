import { FundingNetwork } from '@scripts/cms';

export const funding: FundingNetwork[] = [
	{
		url: 'https://example.com/1',
		name: 'Funding 1',
		index: 0,
		matrix: 'matrix(1,0,0,1,0,0)'
	},
	{
		url: 'https://example.com/2',
		name: 'Funding 2',
		index: 1,
		matrix: 'matrix(0,1,1,0,1,1)'
	}
];
