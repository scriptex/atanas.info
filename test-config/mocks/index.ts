export * from './articles';
export * from './certificates';
export * from './experience';
export * from './lastfm-insights';
export * from './resume';
export * from './slides';
export * from './strengths';
export * from './timeline';
export * from './videos';
