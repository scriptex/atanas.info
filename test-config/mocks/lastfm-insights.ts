import { LastFMInsights } from '@insights/utils';

export const lastFmInsights: LastFMInsights = {
	error: false,
	info: {
		user: {
			age: '39',
			album_count: '1000',
			artist_count: '1234',
			bootstrap: '',
			country: 'Bulgaria',
			gender: 'male',
			image: [],
			name: 'scriptex',
			playcount: '12345',
			playlists: '10',
			realname: 'Atanas',
			registered: {
				'#text': 123456789,
				unixtime: '123456789'
			},
			subscriber: '',
			track_count: '100',
			type: 'user',
			url: 'https://atanas.info'
		}
	},
	topAlbums: [
		{
			artist: 'Test Artist',
			images: [
				{
					'#text': 'https://lastfm.freetls.fastly.net/i/u/34s/2b20f07b88e2de001eeb4ea110d6a061.png',
					size: 'small'
				},
				{
					'#text': 'https://lastfm.freetls.fastly.net/i/u/64s/2b20f07b88e2de001eeb4ea110d6a061.png',
					size: 'medium'
				},
				{
					'#text': 'https://lastfm.freetls.fastly.net/i/u/174s/2b20f07b88e2de001eeb4ea110d6a061.png',
					size: 'large'
				},
				{
					'#text': 'https://lastfm.freetls.fastly.net/i/u/300x300/2b20f07b88e2de001eeb4ea110d6a061.png',
					size: 'extralarge'
				}
			],
			name: 'Test top album'
		}
	],
	updated: 1675168164186,
	weeklyAlbumChart: [
		{
			artist: 'Test Artist',
			images: [
				{
					'#text': 'https://lastfm.freetls.fastly.net/i/u/34s/ade81b221def43c48ba7b364058c13c2.png',
					size: 'small'
				},
				{
					'#text': 'https://lastfm.freetls.fastly.net/i/u/64s/ade81b221def43c48ba7b364058c13c2.png',
					size: 'medium'
				},
				{
					'#text': 'https://lastfm.freetls.fastly.net/i/u/174s/ade81b221def43c48ba7b364058c13c2.png',
					size: 'large'
				},
				{
					'#text': 'https://lastfm.freetls.fastly.net/i/u/300x300/ade81b221def43c48ba7b364058c13c2.png',
					size: 'extralarge'
				}
			],
			name: 'Test weekly album'
		}
	]
};
