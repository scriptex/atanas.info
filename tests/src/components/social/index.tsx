import { LinkedInBadge, SocialMusic } from '@components';
import type { LastFMAlbum } from '@insights/utils';
import { mockFetch, snapshotTest } from '@test-config/helpers';
import { lastFmInsights, recentTracks } from '@test-config/mocks';

mockFetch(() => recentTracks);

snapshotTest(LinkedInBadge);

snapshotTest(() => <SocialMusic data={lastFmInsights} />, undefined, 'SocialMusic');

snapshotTest(() => <SocialMusic data={{ ...lastFmInsights, topAlbums: [] }} />, undefined, 'SocialMusic');

snapshotTest(() => <SocialMusic data={{ ...lastFmInsights, error: true }} />, undefined, 'SocialMusic');

snapshotTest(
	() => <SocialMusic data={{ ...lastFmInsights, topAlbums: [{} as LastFMAlbum] }} />,
	undefined,
	'SocialMusic'
);
