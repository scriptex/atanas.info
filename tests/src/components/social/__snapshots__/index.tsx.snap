// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LinkedInBadge component Should render the LinkedInBadge component 1`] = `
<DocumentFragment>
  <div
    class="linkedin-badge bubbles"
  >
    <div
      class="linkedin-badge__content"
    >
      <img
        alt="Atanas Atanasov smiling dressed in a green t-shirt"
        data-nimg="1"
        decoding="async"
        height="240"
        loading="lazy"
        src="/_next/image?url=%2Fimages%2Ftemp%2Fatanas.jpg&w=640&q=75"
        srcset="/_next/image?url=%2Fimages%2Ftemp%2Fatanas.jpg&w=256&q=75 1x, /_next/image?url=%2Fimages%2Ftemp%2Fatanas.jpg&w=640&q=75 2x"
        style="color: transparent;"
        width="240"
      />
      <h4>
        <a
          href="https://linkedin.com/in/scriptex"
          rel="noopener noreferrer"
          target="_blank"
        >
          ðŸ‘¨ðŸ’» Atanas Atanasov
          <svg
            class="c-svg-external-link"
          >
            <use
              xlink:href="#svg-external-link"
            />
          </svg>
        </a>
      </h4>
      <h5>
        Lead Front-end Engineer / E.ON Home Project @E.ON
      </h5>
      <h5>
        <a
          href="https://www.linkedin.com/company/e-on"
          rel="noopener noreferrer"
          target="_blank"
        >
          E.ON
          <svg
            class="c-svg-external-link"
          >
            <use
              xlink:href="#svg-external-link"
            />
          </svg>
        </a>
         |
        <a
          href="https://www.linkedin.com/school/university-of-economics-varna"
          rel="noopener noreferrer"
          target="_blank"
        >
          University of Economics - Varna
          <svg
            class="c-svg-external-link"
          >
            <use
              xlink:href="#svg-external-link"
            />
          </svg>
        </a>
      </h5>
      <a
        class="c-btn c-btn--small"
        href="https://linkedin.com/in/scriptex"
        rel="noopener noreferrer"
        target="_blank"
      >
        View profile
        <svg
          class="c-svg-external-link"
        >
          <use
            xlink:href="#svg-external-link"
          />
        </svg>
      </a>
    </div>
  </div>
</DocumentFragment>
`;

exports[`LinkedInBadge component Should unmount the LinkedInBadge component 1`] = `<DocumentFragment />`;

exports[`SocialMusic component Should render the SocialMusic component 1`] = `
<DocumentFragment>
  <header
    class="c-section__header"
  >
    <h2>
      <a
        href="https://www.last.fm/user/scriptex"
        rel="noopener noreferrer"
        target="_blank"
      >
        Last.FM
        <svg
          class="c-svg-external-link"
        >
          <use
            xlink:href="#svg-external-link"
          />
        </svg>
      </a>
       statistics
      <small>
        Updated at Mock date
      </small>
    </h2>
  </header>
  <div
    class="c-section--stats"
  >
    <div
      class="c-section__entry c-section__entry--no-background"
    >
      <div
        class="o-shell"
      >
        <ul
          class="c-section__list"
        >
          <li>
            <span>
              Last.FM username:
            </span>
            <strong>
              scriptex
            </strong>
          </li>
          <li>
            <span>
              Total albums:
            </span>
            <strong>
              1000
            </strong>
          </li>
          <li>
            <span>
              Total artists:
            </span>
            <strong>
              1234
            </strong>
          </li>
          <li>
            <span>
              Total plays:
            </span>
            <strong>
              12345
            </strong>
          </li>
          <li>
            <span>
              Total tracks:
            </span>
            <strong>
              100
            </strong>
          </li>
          <li>
            <span>
              Listening since:
            </span>
            <strong>
              Mock date
            </strong>
          </li>
        </ul>
      </div>
    </div>
  </div>
  <div
    class="c-section__entry"
  >
    <div
      class="o-shell"
    >
      <div
        class="o-grid"
      >
        <div
          class="o-grid__item xs-12"
        >
          <h3>
            Top albums for last week:
          </h3>
          <div
            class="carousel"
          >
            <div
              class="carousel__container"
              style="transform: translateZ(857390066042756900px) rotateY(0deg);"
            >
              <div
                class="carousel__slide"
                style="opacity: 1; transform: rotateY(0deg) translateZ(-857390066042756900px);"
              >
                <img
                  alt="Test weekly album"
                  src="https://lastfm.freetls.fastly.net/i/u/174s/ade81b221def43c48ba7b364058c13c2.png"
                />
                <div
                  class="carousel__slide-overlay"
                >
                  <strong>
                    Test weekly album
                  </strong>
                  <span>
                    Test Artist
                  </span>
                </div>
              </div>
            </div>
          </div>
          <div
            class="carousel__controls"
          >
            <button
              class="carousel__control carousel__control--prev"
            >
              Previous
            </button>
            <button
              class="carousel__control carousel__control--next"
            >
              Next
            </button>
          </div>
        </div>
        <div
          class="o-grid__item xs-12"
        >
          <h3>
            Top albums for last month:
          </h3>
          <div
            class="carousel"
          >
            <div
              class="carousel__container"
              style="transform: translateZ(857390066042756900px) rotateY(0deg);"
            >
              <div
                class="carousel__slide"
                style="opacity: 1; transform: rotateY(0deg) translateZ(-857390066042756900px);"
              >
                <img
                  alt="Test top album"
                  src="https://lastfm.freetls.fastly.net/i/u/174s/2b20f07b88e2de001eeb4ea110d6a061.png"
                />
                <div
                  class="carousel__slide-overlay"
                >
                  <strong>
                    Test top album
                  </strong>
                  <span>
                    Test Artist
                  </span>
                </div>
              </div>
            </div>
          </div>
          <div
            class="carousel__controls"
          >
            <button
              class="carousel__control carousel__control--prev"
            >
              Previous
            </button>
            <button
              class="carousel__control carousel__control--next"
            >
              Next
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`SocialMusic component Should render the SocialMusic component 2`] = `
<DocumentFragment>
  <header
    class="c-section__header"
  >
    <h2>
      <a
        href="https://www.last.fm/user/scriptex"
        rel="noopener noreferrer"
        target="_blank"
      >
        Last.FM
        <svg
          class="c-svg-external-link"
        >
          <use
            xlink:href="#svg-external-link"
          />
        </svg>
      </a>
       statistics
      <small>
        Updated at Mock date
      </small>
    </h2>
  </header>
  <div
    class="c-section--stats"
  >
    <div
      class="c-section__entry c-section__entry--no-background"
    >
      <div
        class="o-shell"
      >
        <ul
          class="c-section__list"
        >
          <li>
            <span>
              Last.FM username:
            </span>
            <strong>
              scriptex
            </strong>
          </li>
          <li>
            <span>
              Total albums:
            </span>
            <strong>
              1000
            </strong>
          </li>
          <li>
            <span>
              Total artists:
            </span>
            <strong>
              1234
            </strong>
          </li>
          <li>
            <span>
              Total plays:
            </span>
            <strong>
              12345
            </strong>
          </li>
          <li>
            <span>
              Total tracks:
            </span>
            <strong>
              100
            </strong>
          </li>
          <li>
            <span>
              Listening since:
            </span>
            <strong>
              Mock date
            </strong>
          </li>
        </ul>
      </div>
    </div>
  </div>
  <div
    class="c-section__entry"
  >
    <div
      class="o-shell"
    >
      <div
        class="o-grid"
      >
        <div
          class="o-grid__item xs-12"
        >
          <h3>
            Top albums for last week:
          </h3>
          <div
            class="carousel"
          >
            <div
              class="carousel__container"
              style="transform: translateZ(857390066042756900px) rotateY(0deg);"
            >
              <div
                class="carousel__slide"
                style="opacity: 1; transform: rotateY(0deg) translateZ(-857390066042756900px);"
              >
                <img
                  alt="Test weekly album"
                  src="https://lastfm.freetls.fastly.net/i/u/174s/ade81b221def43c48ba7b364058c13c2.png"
                />
                <div
                  class="carousel__slide-overlay"
                >
                  <strong>
                    Test weekly album
                  </strong>
                  <span>
                    Test Artist
                  </span>
                </div>
              </div>
            </div>
          </div>
          <div
            class="carousel__controls"
          >
            <button
              class="carousel__control carousel__control--prev"
            >
              Previous
            </button>
            <button
              class="carousel__control carousel__control--next"
            >
              Next
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`SocialMusic component Should render the SocialMusic component 3`] = `<DocumentFragment />`;

exports[`SocialMusic component Should render the SocialMusic component 4`] = `
<DocumentFragment>
  <header
    class="c-section__header"
  >
    <h2>
      <a
        href="https://www.last.fm/user/scriptex"
        rel="noopener noreferrer"
        target="_blank"
      >
        Last.FM
        <svg
          class="c-svg-external-link"
        >
          <use
            xlink:href="#svg-external-link"
          />
        </svg>
      </a>
       statistics
      <small>
        Updated at Mock date
      </small>
    </h2>
  </header>
  <div
    class="c-section--stats"
  >
    <div
      class="c-section__entry c-section__entry--no-background"
    >
      <div
        class="o-shell"
      >
        <ul
          class="c-section__list"
        >
          <li>
            <span>
              Last.FM username:
            </span>
            <strong>
              scriptex
            </strong>
          </li>
          <li>
            <span>
              Total albums:
            </span>
            <strong>
              1000
            </strong>
          </li>
          <li>
            <span>
              Total artists:
            </span>
            <strong>
              1234
            </strong>
          </li>
          <li>
            <span>
              Total plays:
            </span>
            <strong>
              12345
            </strong>
          </li>
          <li>
            <span>
              Total tracks:
            </span>
            <strong>
              100
            </strong>
          </li>
          <li>
            <span>
              Listening since:
            </span>
            <strong>
              Mock date
            </strong>
          </li>
        </ul>
      </div>
    </div>
  </div>
  <div
    class="c-section__entry"
  >
    <div
      class="o-shell"
    >
      <div
        class="o-grid"
      >
        <div
          class="o-grid__item xs-12"
        >
          <h3>
            Top albums for last week:
          </h3>
          <div
            class="carousel"
          >
            <div
              class="carousel__container"
              style="transform: translateZ(857390066042756900px) rotateY(0deg);"
            >
              <div
                class="carousel__slide"
                style="opacity: 1; transform: rotateY(0deg) translateZ(-857390066042756900px);"
              >
                <img
                  alt="Test weekly album"
                  src="https://lastfm.freetls.fastly.net/i/u/174s/ade81b221def43c48ba7b364058c13c2.png"
                />
                <div
                  class="carousel__slide-overlay"
                >
                  <strong>
                    Test weekly album
                  </strong>
                  <span>
                    Test Artist
                  </span>
                </div>
              </div>
            </div>
          </div>
          <div
            class="carousel__controls"
          >
            <button
              class="carousel__control carousel__control--prev"
            >
              Previous
            </button>
            <button
              class="carousel__control carousel__control--next"
            >
              Next
            </button>
          </div>
        </div>
        <div
          class="o-grid__item xs-12"
        >
          <h3>
            Top albums for last month:
          </h3>
          <div
            class="carousel"
          >
            <div
              class="carousel__container"
              style="transform: translateZ(NaNpx) rotateY(NaNdeg);"
            />
          </div>
          <div
            class="carousel__controls"
          >
            <button
              class="carousel__control carousel__control--prev"
            >
              Previous
            </button>
            <button
              class="carousel__control carousel__control--next"
            >
              Next
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`SocialMusic component Should unmount the SocialMusic component 1`] = `<DocumentFragment />`;

exports[`SocialMusic component Should unmount the SocialMusic component 2`] = `<DocumentFragment />`;

exports[`SocialMusic component Should unmount the SocialMusic component 3`] = `<DocumentFragment />`;

exports[`SocialMusic component Should unmount the SocialMusic component 4`] = `<DocumentFragment />`;
